### DISCUSSION TITLE
Qestion bout nd.waitall()

### DISCUSSION THREAD

**Discussion by Jessespace** [apache/incubator-mxnet#14220] [Q&A]
I want to test the inference time about yolov3.I I wrote two versions.like bellow:

1.`tic = time.time()
class_IDs, scores, bounding_boxs = net(x)
toc = time.time()
print("Inference time is :",toc-tic)`

2.`tic = time.time()
class_IDs, scores, bounding_boxs = net(x)
**nd.waitall()**
toc = time.time()
print("Inference time is :",toc-tic)`

The results of the two versions vary greatly.Which one is right? and Why?

[a comment by an automated account is not shown]

**Comment by frankfliu** [comment]
@mxnet-label-bot add [Question]

**Comment by wkcn** [comment]
2 is right.
MXNet computes operators asynchronously, so it is necessary to call 'nd.waitall()' to wait for all computation over.

**Comment by frankfliu** [comment]
@Jessespace 
MXNet ndarray use lazy evaluation, the previous operations may not finished yet. For benchmark purpose, you should use nd.wait_to_read() or nd.waitall(). 
See: https://mxnet.incubator.apache.org/api/python/ndarray/ndarray.html#mxnet.ndarray.NDArray.wait_to_read
And: https://mxnet.incubator.apache.org/api/python/ndarray/ndarray.html#mxnet.ndarray.waitall

**Comment by Jessespace** [comment]
> @Jessespace
> MXNet ndarray use lazy evaluation, the previous operations may not finished yet. For benchmark purpose, you should use nd.wait_to_read() or nd.waitall().
> See: https://mxnet.incubator.apache.org/api/python/ndarray/ndarray.html#mxnet.ndarray.NDArray.wait_to_read
> And: https://mxnet.incubator.apache.org/api/python/ndarray/ndarray.html#mxnet.ndarray.waitall

I'm still not clear about this problem.I use **nd.waitall()** to test a single image.
But I got 0.49 FPS.Please help me figure out  the problem.

```
path = os.path.join(read_dir, '000204.jpg')
tic = time.time()
img_path = os.path.join(read_dir, path)
x, img = data.transforms.presets.yolo.load_test(img_path, short=shape,max_size=shape)
x = x.as_in_context(ctx)
time_loadimage = time.time() - tic

class_IDs, scores, bounding_boxs = net(x)
nd.waitall()
all = time.time() - tic

print("every picture time cost:", (all - time_loadimage))
print("FPS:", 1.0 / (all - time_loadimage))
```

**Comment by stephenrawls** [comment]
It is unclear from your question. Is the problem that you think it is slower than it should be?

If you are using the GPU, just FYI there is a lot of overhead for the first operation you perform on the GPU. So for benchmarking purposes you need to "warm up" the GPU by processing at least one item on the GPU, and then start timing for real after the initialization overhead is finished.

**Comment by thomelane** [comment]
You'll want to start 'time_loadimage' after the load and context change has finished: remember that MXNet operations are asynchronous. And it's better to use `wait_to_read()` instead of `waitall()` if you're wanting to time certain operations. And even better still you should check out the [MXNet Profiler](https://mxnet.incubator.apache.org/versions/master/tutorials/python/profiler.html) instead of writing your own profiling code.

```
path = os.path.join(read_dir, '000204.jpg')
tic = time.time()
img_path = os.path.join(read_dir, path)
x, img = data.transforms.presets.yolo.load_test(img_path, short=shape,max_size=shape)
x = x.as_in_context(ctx)
# block until data load and context switch have completed
x.wait_to_read()
time_loadimage = time.time() - tic

class_IDs, scores, bounding_boxs = net(x)
# block until net has completed
bounding_boxs.wait_to_read()
all = time.time() - tic

print("every picture time cost:", (all - time_loadimage))
print("FPS:", 1.0 / (all - time_loadimage))
```

**Comment by Jessespace** [comment]
> You'll want to start 'time_loadimage' after the load and context change has finished: remember that MXNet operations are asynchronous. And it's better to use `wait_to_read()` instead of `waitall()` if you're wanting to time certain operations. And even better still you should check out the [MXNet Profiler](https://mxnet.incubator.apache.org/versions/master/tutorials/python/profiler.html) instead of writing your own profiling code.
> 
> ```
> path = os.path.join(read_dir, '000204.jpg')
> tic = time.time()
> img_path = os.path.join(read_dir, path)
> x, img = data.transforms.presets.yolo.load_test(img_path, short=shape,max_size=shape)
> x = x.as_in_context(ctx)
> # block until data load and context switch have completed
> x.wait_to_read()
> time_loadimage = time.time() - tic
> 
> class_IDs, scores, bounding_boxs = net(x)
> # block until net has completed
> bounding_boxs.wait_to_read()
> all = time.time() - tic
> 
> print("every picture time cost:", (all - time_loadimage))
> print("FPS:", 1.0 / (all - time_loadimage))
> ```



> You'll want to start 'time_loadimage' after the load and context change has finished: remember that MXNet operations are asynchronous. And it's better to use `wait_to_read()` instead of `waitall()` if you're wanting to time certain operations. And even better still you should check out the [MXNet Profiler](https://mxnet.incubator.apache.org/versions/master/tutorials/python/profiler.html) instead of writing your own profiling code.
> 
> ```
> path = os.path.join(read_dir, '000204.jpg')
> tic = time.time()
> img_path = os.path.join(read_dir, path)
> x, img = data.transforms.presets.yolo.load_test(img_path, short=shape,max_size=shape)
> x = x.as_in_context(ctx)
> # block until data load and context switch have completed
> x.wait_to_read()
> time_loadimage = time.time() - tic
> 
> class_IDs, scores, bounding_boxs = net(x)
> # block until net has completed
> bounding_boxs.wait_to_read()
> all = time.time() - tic
> 
> print("every picture time cost:", (all - time_loadimage))
> print("FPS:", 1.0 / (all - time_loadimage))
> ```

I followed your advice and tried to use **profile**. Still some confusion about some places.

**code:**
profiler.set_state('run')
tic = time.time()

x, img = data.transforms.presets.yolo.load_test(img_path, short=shape)
x = x.as_in_context(ctx)
class_IDs, scores, bounding_boxs = net(x)

bounding_boxs.wait_to_read()
all = time.time() - tic
profiler.set_state('stop')
print("time:", all)
print(profiler.dumps())`
    

**Result:**
   1. time: 5.4537177085876465
   2.  [profile result：](https://github.com/Jessespace/MXNet-Learning/blob/master/profile.png)

I add  all the **MXNET_C_API time** ,it is about the 5.4s which similar to 5.45s using **tic toc**. But I can't understand **Device Storage time** and **operator time** ,they take too long.

**Comment by Hao-Tju** [comment]
> > You'll want to start 'time_loadimage' after the load and context change has finished: remember that MXNet operations are asynchronous. And it's better to use `wait_to_read()` instead of `waitall()` if you're wanting to time certain operations. And even better still you should check out the [MXNet Profiler](https://mxnet.incubator.apache.org/versions/master/tutorials/python/profiler.html) instead of writing your own profiling code.
> > ```
> > path = os.path.join(read_dir, '000204.jpg')
> > tic = time.time()
> > img_path = os.path.join(read_dir, path)
> > x, img = data.transforms.presets.yolo.load_test(img_path, short=shape,max_size=shape)
> > x = x.as_in_context(ctx)
> > # block until data load and context switch have completed
> > x.wait_to_read()
> > time_loadimage = time.time() - tic
> > 
> > class_IDs, scores, bounding_boxs = net(x)
> > # block until net has completed
> > bounding_boxs.wait_to_read()
> > all = time.time() - tic
> > 
> > print("every picture time cost:", (all - time_loadimage))
> > print("FPS:", 1.0 / (all - time_loadimage))
> > ```
> 
> > You'll want to start 'time_loadimage' after the load and context change has finished: remember that MXNet operations are asynchronous. And it's better to use `wait_to_read()` instead of `waitall()` if you're wanting to time certain operations. And even better still you should check out the [MXNet Profiler](https://mxnet.incubator.apache.org/versions/master/tutorials/python/profiler.html) instead of writing your own profiling code.
> > ```
> > path = os.path.join(read_dir, '000204.jpg')
> > tic = time.time()
> > img_path = os.path.join(read_dir, path)
> > x, img = data.transforms.presets.yolo.load_test(img_path, short=shape,max_size=shape)
> > x = x.as_in_context(ctx)
> > # block until data load and context switch have completed
> > x.wait_to_read()
> > time_loadimage = time.time() - tic
> > 
> > class_IDs, scores, bounding_boxs = net(x)
> > # block until net has completed
> > bounding_boxs.wait_to_read()
> > all = time.time() - tic
> > 
> > print("every picture time cost:", (all - time_loadimage))
> > print("FPS:", 1.0 / (all - time_loadimage))
> > ```
> 
> I followed your advice and tried to use **profile**. Still some confusion about some places.
> 
> **code:**
> profiler.set_state('run')
> tic = time.time()
> 
> x, img = data.transforms.presets.yolo.load_test(img_path, short=shape)
> x = x.as_in_context(ctx)
> class_IDs, scores, bounding_boxs = net(x)
> 
> bounding_boxs.wait_to_read()
> all = time.time() - tic
> profiler.set_state('stop')
> print("time:", all)
> print(profiler.dumps())`
> 
> **Result:**
> 
> 1. time: 5.4537177085876465
> 2. [profile result：](https://github.com/Jessespace/MXNet-Learning/blob/master/profile.png)
> 
> I add all the **MXNET_C_API time** ,it is about the 5.4s which similar to 5.45s using **tic toc**. But I can't understand **Device Storage time** and **operator time** ,they take too long.

I recommend that you should not use both the mxnet profiler and the time function at the same time. MXNet profiler will incur additional overhead and result in a wrong time result.
